Remove support for python 3.8

kratman opened the pull request:
# Description

Removes support for python 3.8

Fixes #3390

## Type of change

This is a breaking change since it reduces support for a python version. On the plus side it reduces the number of CI jobs.

# Key checklist:

- [x] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [x] All tests pass: `$ python run-tests.py --all` (or `$ nox -s tests`)
- [x] The documentation builds: `$ python run-tests.py --doctest` (or `$ nox -s doctests`)

You can run integration tests, unit tests, and doctests together at once, using `$ python run-tests.py --quick` (or `$ nox -s quick`).

## Further checks:

- [x] Code is commented, particularly in hard-to-understand areas
- [x] Tests added that prove fix is effective or that feature works

kratman commented:
I merged #3962 into this one to get the tests running faster

agriyakhetarpal commented:
Also, I remember that there were some notes that we had added about the JAX solver not being available for Python 3.8, we should be able to remove those too now (can't find where exactly they were in the installation docs)

agriyakhetarpal commented:
Found them, they are in `gnu-linux-mac.rst` and `windows.rst` under the "Optional – JaxSolver" headings

agriyakhetarpal commented on .github/workflows/benchmark_on_push.yml:
```diff
@@ -13,10 +13,10 @@ jobs:
     steps:
       - uses: actions/checkout@v4
-      - name: Set up Python 3.8
+      - name: Set up Python 3.11
```
```suggestion
      - name: Set up Python 3.12
```

Can we bump this to Python 3.12, it might be marginally faster and all dependencies support it?

agriyakhetarpal commented on .github/workflows/benchmark_on_push.yml:
```diff
@@ -13,10 +13,10 @@ jobs:
         uses: actions/setup-python@v5
         with:
-          python-version: 3.8
+          python-version: 3.11
```
```suggestion
          python-version: 3.12
```

agriyakhetarpal commented on .github/workflows/periodic_benchmarks.yml:
```diff
@@ -21,10 +21,10 @@ jobs:
       - uses: actions/checkout@v4
 
-      - name: Set up Python 3.8
+      - name: Set up Python 3.11
```
```suggestion
      - name: Set up Python 3.12
```

agriyakhetarpal commented on .github/workflows/periodic_benchmarks.yml:
```diff
@@ -21,10 +21,10 @@ jobs:
         uses: actions/setup-python@v5
         with:
-          python-version: 3.8
+          python-version: 3.11
```
```suggestion
          python-version: 3.12
```

agriyakhetarpal commented on .github/workflows/periodic_benchmarks.yml:
```diff
@@ -58,10 +58,10 @@ jobs:
     runs-on: ubuntu-latest
     steps:
-      - name: Set up Python 3.8
+      - name: Set up Python 3.11
```
```suggestion
      - name: Set up Python 3.12
```

agriyakhetarpal commented on .github/workflows/periodic_benchmarks.yml:
```diff
@@ -58,10 +58,10 @@ jobs:
         uses: actions/setup-python@v5
         with:
-          python-version: 3.8
+          python-version: 3.11
```
```suggestion
          python-version: 3.12
```

agriyakhetarpal commented on .github/workflows/publish_pypi.yml:
```diff
@@ -35,7 +35,7 @@ jobs:
       - uses: actions/checkout@v4
       - uses: actions/setup-python@v5
         with:
-          python-version: 3.8
+          python-version: 3.11
```
I think we can remove `actions/setup-python` from all of the wheel jobs and use the [`cibuildwheel` action](https://github.com/pypa/cibuildwheel/blob/main/action.yml) instead which does it for us already (also `cibuildwheel` installs its own Python interpreters). Please feel free to do it here or in a separate PR, it's mostly cleanup though

kratman replied:
Yeah lets make that one a different PR

agriyakhetarpal commented on .github/workflows/publish_pypi.yml:
```diff
@@ -107,7 +107,7 @@ jobs:
       - uses: actions/setup-python@v5
         name: Set up Python
         with:
-          python-version: 3.8
+          python-version: 3.11
```
Same as above

agriyakhetarpal commented on .github/workflows/run_benchmarks_over_history.yml:
```diff
@@ -23,10 +23,10 @@ jobs:
     steps:
       - uses: actions/checkout@v4
-      - name: Set up Python 3.8
+      - name: Set up Python 3.11
```
```suggestion
      - name: Set up Python 3.12
```

same as above

agriyakhetarpal commented on .github/workflows/run_benchmarks_over_history.yml:
```diff
@@ -23,10 +23,10 @@ jobs:
         uses: actions/setup-python@v5
         with:
-          python-version: 3.8
+          python-version: 3.11
```
```suggestion
          python-version: 3.12
```

agriyakhetarpal commented on .github/workflows/run_benchmarks_over_history.yml:
```diff
@@ -53,10 +53,10 @@ jobs:
     runs-on: ubuntu-latest
     steps:
-      - name: Set up Python 3.8
+      - name: Set up Python 3.11
```
```suggestion
      - name: Set up Python 3.12
```

agriyakhetarpal commented on .github/workflows/run_benchmarks_over_history.yml:
```diff
@@ -53,10 +53,10 @@ jobs:
         uses: actions/setup-python@v5
         with:
-          python-version: 3.8
+          python-version: 3.11
```
```suggestion
          python-version: 3.12
```

agriyakhetarpal commented on .github/workflows/run_periodic_tests.yml:
```diff
@@ -92,7 +92,7 @@ jobs:
 
-      - name: Run unit tests for GNU/Linux with Python 3.8, 3.9, 3.10, and 3.12; and for macOS and Windows with all Python versions
+      - name: Run unit tests for GNU/Linux, macOS, and Windows with all Python versions
         if: (matrix.os == 'ubuntu-latest' && matrix.python-version != 3.11) || (matrix.os != 'ubuntu-latest')
```
Now we should be able to run coverage on 3.12

kratman replied:
Lets do that elsewhere

agriyakhetarpal commented on .github/workflows/update_version.yml:
```diff
@@ -44,7 +44,7 @@ jobs:
         uses: actions/setup-python@v5
         with:
-          python-version: 3.8
+          python-version: 3.11
```
```suggestion
          python-version: '3.x'
```

This workflow isn't doing a lot, this should be better?

kratman replied:
I say we do things like that in a different change, I just want this one to cleanly remove 3.8
```suggestion
          python-version: 3.12
```

agriyakhetarpal reviewed (approved):
Thanks! I'm not aware of the typing stuff but it looks okay to me since `pre-commit` is passing.

kratman commented:
> Thanks! I'm not aware of the typing stuff but it looks okay to me since `pre-commit` is passing.

That was done by pre-commit. I guess it detected that I removed 3.8 support and switched a bunch of stuff
